homeassistant/components/tplink_lte.py

```diff
@@ -0,0 +1,109 @@
+"""
+Support for TP-Link LTE modems.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/tplink_lte/
+"""
+import asyncio
+import logging
+from datetime import timedelta
+
+import aiohttp
+import attr
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_HOST, CONF_PASSWORD, EVENT_HOMEASSISTANT_STOP)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.aiohttp_client import async_create_clientsession
+
+REQUIREMENTS = ['tp-connected==0.0.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=10)
+
+DOMAIN = 'tplink_lte'
+DATA_KEY = 'tplink_lte'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+    })])
+}, extra=vol.ALLOW_EXTRA)
+
+
+@attr.s
+class ModemData:
+    """Class for modem state."""
+
+    modem = attr.ib()
+
+
+@attr.s
+class LTEData:
+    """Shared state."""
+
+    websession = attr.ib()
+    modem_data = attr.ib(init=False, factory=dict)
+
+    def get_modem_data(self, config):
+        """Get the requested or the only modem_data value."""
+        if CONF_HOST in config:
+            return self.modem_data.get(config[CONF_HOST])
+        if len(self.modem_data) == 1:
+            return next(iter(self.modem_data.values()))
+
+        return None
+
+
+async def async_setup(hass, config):
+    """Set up TP-Link LTE component."""
+    if DATA_KEY not in hass.data:
+        websession = async_create_clientsession(
+            hass, cookie_jar=aiohttp.CookieJar(unsafe=True))
+        hass.data[DATA_KEY] = LTEData(websession)
+
+    tasks = [_setup_lte(hass, conf) for conf in config.get(DOMAIN, [])]
```

Comment:
Move default list to config schema.

Reply:
Could you please be more specific or give me some indication to other components that do this?

Reply:
I think we can just use `dict[key]` here instead. The domain key will be required to set up the component so we shouldn't need any defaults here or in the schema. Sorry  for my confusion.

Reply:
I was mistaken. We don't yet validate that the `'tplink_lte'` key for the component is present in the config. Since that is required for a functioning integration, we should check that early and fail component setup if the config isn't ok.

Maybe it would make sense to move the whole config to the component and set up the notify platform via discovery instead. This will releave the user from having to juggle multiple config sections.

Reply:
For the `netgear_lte` component I opted out of discovery because I saw no way to set things like notify `name` and sensor `scan_interval`. Was I mistaken?

Reply:
We can send whatever config we want as discovery_info. But if the platform is using home assistant automatic polling we can't change the scan interval via discovery_info.

Reply:
I'm not sure about the relative configuration to enable the discovery, my first thought was something like this:

```yaml
tplink_lte:
  - host: 192.168.5.1
    password: *********
    notifies:
      - name: sms1
        target: "+3955555555"
      - name: sms2
        target: "+3955555555"
```

but `notifies` seems to be not standard, could you please give me an advice on this?

Reply:
`devices` maybe?

Reply:
I found the `konnected` component as a good example of this , IMO. I think I would use `devices` with another nested field like `targets`. For example:

```yaml
tplink_lte:
  - host: 192.168.5.1
    password: *********
    devices:
      targets:
        - name: sms1
          target: "+3955555555"
        - name: sms2
          target: "+3955555555"
```

My only concern is that `targets` seems to me too general, I would prefer something like `notifies` or some other alternative.

Reply:
`notifies` is not plural of a noun. `targets` matches the sub key `target` so sounds good to me.

Reply:
I'm going to update the docs when this conversation is resolved

Reply:
I think `devices` is odd here, isn't the `host` the device?

Reply:
@amelchio you're right, the host is the physical device but I'm not sure about the right solution here, something like this is enough?

```yaml
tplink_lte:
  - host: 192.168.5.1
     password: *********
     targets:
       - name: sms1
         target: "+3955555555"
       - name: sms2
         target: "+3955555555"
```

Reply:
I am not sure myself. But once we decide, I want to do this for netgear_lte as well :)

I was thinking about using the component name of the platform that we setup ... so `notify` and at least netgear_lte will also need configuration for `sensor`...

```yaml
netgear_lte:
  - host: 192.168.5.1
    password: *********
    notify:
      - name: sms1
        target: "+3955555555"
      - name: sms2
        target: "+3955555555"
    sensor:
      - usage
      - sms

```

This has some precedence, for example with the Cast component `media_player` configuration. But I am starting to wonder whether this is actually an architectural question. It would be really good to have guidelines for this quite common situation.

BTW @andtos90, it's awesome to see you keep positive during that rather tough review 👍

Reply:
I like this solution but should it be `sensors` instead of `sensor`? Just to know for future developments.

Your team is doing a great work and in the process I'm learning a lot, thanks for your work :smile:

Reply:
I am not sure and there are no rules :-/

We create `sensor` entities, that's why I picked singular. I agree that it looks a bit odd if one reads it just as yaml.